Do not emit unused fields in RDG source

martincostello opened the pull request:
# Do not emit unused fields in RDG source

Elide private fields that are not used by emitted RDG code.

## Description

Do not emit unused private fields for HTTP verbs that are not used by any of the user-code endpoints. 

Fixes #48381.

mitchdenny commented:
Thanks @martincostello, this looks good and is well covered by our existing test cases. I'm approving, but also tagging @captainsafia just in case there is an issue I've missed.

captainsafia commented:
@martincostello @mitchdenny Apologies for the delay! 

I was trying to form a more coherent note of my feedback here.

The gist of my reservations is that it feels rather strange for us to use a completely different strategy for emit for this static fields compared to what we have for the existing helper types and methods. Typically, we'd use the EmitterContext to set state for types/methods that need to be emitted as we render their invocations. It would be worthwhile to explore doing this instead of the mapping out the verbs in `EmitVerbs` and see if we can have a discrete `helperFields` pipeline for those fields.

martincostello commented:
No problem - I wasn't sure on what the _best_ way was, so went with an initial approach that wouldn't churn the code for the generator too much.

If you can point me at roughly what you're referring to for the context and the pipeline I'll look at changing the approach once #48555 is merged as otherwise I think it'll be a massive merge conflict nightmare.

captainsafia commented on src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs:
```diff
@@ -161,15 +163,22 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
                         codeWriter.WriteLine($"{SymbolDisplay.FormatLiteral("{*path:nonfile}", true)},");
                     }
+
+                    (var verbSymbol, var verbHttp) = endpoint.EmitVerb();
```
Instead of returning a tuple, we can have `EmitVerb` set state in the `Endpoint.EmitterContext` to indicate which verbs need to have fields emitted for. We can process that state in a new `helperFields` pipeline. That should hopefully flow with our current pattern and make it easy to rebase on the interceptors branch.

mitchdenny commented:
@martincostello this is looking good. Looking at the Helix test failures it looks like the baselines just need to be regenerated.

martincostello committed:
Do not emit unused fields in RDG source

Do not emit unused private fields for HTTP verbs that are not used by any of the user-code endpoints.
Resolves #48381.

mitchdenny commented:
/app run

captainsafia commented on src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs:
```diff
@@ -166,7 +167,12 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
+                    .Where(verb => verb is not null)
+                    .ToImmutableHashSet();
+
+                return (stringWriter.ToString(), verbs);
```
Can we process the verbs in a separate incremental pipeline?

```
var helperFields = endpoints.Select(endpoint => ...);
```

We might even be able to implement a custom comparer in `WithComparer` so we only pick up unique instances of the HTTP method. Thoughts?

martincostello replied:
I'll take a look 👍

martincostello committed:
Move verbs to incremental pipeline

Use another incremental pipeline to produce the HTTP verbs.

martincostello committed:
Add custom comparer

Add custom comparer for endpoints by their HTTP method.

captainsafia reviewed (approved):
🤩
